### ISSUE TITLE
Moving components into packages of a monorepo breaks the babelifiedMessages

### ISSUE THREAD

**Issue by TomFreudenberg** [Avansai/next-multilingual#38]
Hi @nbouvrette Nicolas

have you ever done some work in a monorepo?

I have changed our codebase and some of the translated components are moved into a packages folder in second workplace.

If you use inside a component from __packages/components__:

```js
import { getMessages } from 'next-multilingual/messages';
```

it will break the main process due to:

```js
function getBabelifiedMessages(sourceFilePath) {
    var parsedSourceFile = path.parse(sourceFilePath);
    var sourceFileDirectoryPath = parsedSourceFile.dir;
...
```

while `sourceFileDirectoryPath` is __relative__ (missing root).

I just made a quick fix into the node_modules js file

```js
if (sourceFileDirectoryPath.startsWith('Users/')) sourceFileDirectoryPath = '/' + sourceFileDirectoryPath;
```

which does the work now on my machine.

I am thinking if a path translator should be added in the plugin configuration like `imports` from webpack or `paths` from tsconfig.json

like:

```json
  pluginConfiguration: [
    "next-multilingual-alternate/messages/babel-plugin",
    {
      applicationId: appId,
      ...,
      paths: {
        "Users/"; "/Users/"
      }
    }
  ]
```

or if it just should check if __relative__ path exists and if not try as __full__ path.

Thanks for feedback and hopefully you are fine

Tom

**Comment by nbouvrette** [comment]
Hi @TomFreudenberg 

Long time no talk. I don't have much experience playing with mono-repos, but it also looks like you are using an old version.

Last summer I moved most of the Babel plugin logic into this new package: https://github.com/Avansai/messages-modules

I think if we would need to support your use case we would probably need to modify this package as well. I'm not too sure where the code you provided was used and also not sure I fully understand what is going on in your scenario.

It might be easier to jump on a call like we did last time to save on some back and forth if this can be an option for you.

**Comment by TomFreudenberg** [comment]
Yes @nbouvrette Nicolas :-) long time ago but still using your component :-)

I use my fork and be aligned with your version v0.12.0 from June 2022.

I will try to apply our needs to your current stable and come back to you.

Have you already got the new NEXTjs 13 features from src/app ?

Looks pretty interesting what is coming up there.

Will give you more feedback later

Tom

**Comment by nbouvrette** [comment]
I'm about to release version 3 which will support dynamic catch-all routes (probably tomorrow). There will be a few breaking changes that I will document.

Regarding Next 13 it's on my to-do list to check. If I understood correctly from the conference yesterday the routing patterns should stay the same which is good news for us!

**Comment by TomFreudenberg** [comment]
> Regarding Next 13 it's on my to-do list to check. If I understood correctly from the conference yesterday the routing patterns should stay the same which is good news for us!

yeah

> I'm about to release version 3 which will support dynamic catch-all routes (probably tomorrow). There will be a few breaking changes that I will document.

ok thanks for the info. then i will wait for your update and apply my work on top

currently i have enough to do with moving our work into turbo monorepo ;-)

**Comment by nbouvrette** [comment]
Do you know if Truborepo forces you to use SWC? 

This is a big challenge for next-multilingual because it looks like we would need to rewrite our plugin in Rust. That part is not too bad but the problem is that if we would want the same level of details we are getting with our own https://github.com/Avansai/properties-file package, we would also need to rewrite it in Rust...

**Comment by TomFreudenberg** [comment]
As far as I can get it by now, babel ist still available

**Comment by nbouvrette** [comment]
Great news. Looking forward to try this myself shortly as well

**Comment by TomFreudenberg** [comment]
Hey @nbouvrette Nicolas

As I see, your update is only - so I will start on that?

**Comment by nbouvrette** [comment]
Yes, I don't foresee significant changes any time soon unless we find unexpected bugs with the new version but the +800 tests should have hopefully caught most of them :)

**Comment by TomFreudenberg** [comment]
Hi @nbouvrette Nicolas

I was currently only able to align your release 1.0.7

Changes due to your new packages `message-modules` and `properties-file` will break my changes while I still use yaml files for the translation files.

I guess in case of your strict design decisions it won't be possible to do it as a translation-file package which can handle more than properties file.

I will come up to you when digging more deeper into.

Cheers
Tom

**Comment by nbouvrette** [comment]
One reason I wrote the new `properties-file` package was to get more insights into possible issues with localizable assets.

As we discussed in the past (I think) my main concern with yaml files is that you can use them for more than just key/values which might cause issues during the translation process.

Nobody else asked for other file support so far, which is either because there isn't a lot of serious users of the package around (even tho we hit 150 stars), or the current solution meet most needs.

I don't have a ton of experience personally supporting yaml files with commercial TMSes but  I do recall various issues related to parsing some files so I am still hesitant to consider official support for them.

**Comment by TomFreudenberg** [comment]
Yeah I know and respect your concerns and arguments.

Mostly my biggest issue still with the current properties files is the force / need of "app_id.segment.key" instead of just having "key".

I mean we have discussed this also already - applicationId should not a must on properties file and segment is auto-adjustable to source-filename.

While this may not valid or good for blogs and lot of written international content, but it is the best solution on app development for us.

I check out the current state and come back to you ;-)

**Comment by nbouvrette** [comment]
OK cool and I still had  a few thing in our backlog that we never got to. Not sure if they would help or not at this point but keep me posted!

**Comment by nbouvrette** [comment]
Yes, this error could happen if you don't configure your Webpack handler correctly. There is a weird issue where Webpack doesn't handle imports using the `node:` prefix that has been available for a while and that is also the recommended way to import Node-native modules now.

I presume you are using the advanced config? Can you make sure your Webpack config is set correctly based on the info provided here: https://github.com/Avansai/next-multilingual#%EF%B8%8F-config-advanced-config

**Comment by TomFreudenberg** [comment]
Hi Nicolas @nbouvrette 

thanks for the hint, I got it working - I missed the requirement for webpack configuration!

Now I come back to your .properties decision:

1) AppIdentifier

This may make sense if I have to translate a lot of applications at the same time.

I would say - most developers are working and translating on ONE application at the same time. Also I assume, mostly they will translate the (few) app strings by themself or in the small team.

So for that, repeating the "AppIdentifier" each line in each .properties file is just annoying.

2) Section

Again, this may make sense if I have to translate a lot of applications at the same time.

With same assumption as 1), the "section" identifier is just some (random) unique key to make sure that the correct translations are used on each source. But, the unique key of each source file is the "path/filename" of themself.

So "section" could be build automatically by "path/filename" from source file.

---

Summarize 1) and 2): ApplicationId and Section should not be required always in the .properties file. For small teams and apps it is versus the DRY-rules

---

3) Key

a) Your key is required to be only alphanumeric and numbers from 1-50 chars. 

b) If key is not found in .properties file, NOTHING is shown in the app but only a warning on console.

This decision a) only allows to have a kind of "cryptical" ids in your app while developing and enforce you to directly fill in the translation files as well.

While using hot reloading features and working on the app, there are moments when nothing is shown in the app due to b)

I can't get the necessarity for your decision 3) at all. Maybe it is that you need to enforce your developers on your development projects but on our - and I would say again also on any small teams and apps - projects it is just annoying.

---

As already said on our phone call, there are some requirements for us to use next-multilingual:

1) ApplicationId is not required in .properties file
2) "section" will be generated automatically from "path/filename" and is not required in .properties file
3) "keys" may contain normal text
4) If a "key" is not found in the translation file, the "key" itself will be shown

Last, as you can get now from our above requirements 1), 2), 3) also the .properties file type is not necessary in all cases. Having keys in a full text manner, then also or even better YAML file type can handle the translation files:

```yaml
Access denied: Kein Zugang
Cancel: Zurück

You are not signed in!: Sie sind nicht angemeldet!
You do not have proper rights!: Sie haben keine ausreichenden Rechte!
```

or even lines with line breaks

```yaml
Approval: Posteingang prüfen
Cancel: Abbruch

A sign-in link has been sent to your email address.: >-
  Ein Link zur Anmeldung wurde<br/>
  an Ihre Email‑Adresse versendet.
```

For us, YAML file type has a high visibility for text content.

---

As said in our phone call, I would like to help and work on OPTIONAL settings to just ALLOW our requirements in addition by the "Config" and not for default. I respect your decisions but they don't fit in any case ;-)

Looking forward to your feedback.

Cheers
Tom

**Comment by nbouvrette** [comment]
I went through your comments, and unless I am misunderstanding how you deal with your translation on your projects, it really sounds like you operate in teams that are autonomous. Basically your team both handles the development and the translation of your application. In these conditions, these constraints can seem annoying like you said but let me try to explain why they are there.

### 1) Key prefixes

Regarding the `applicationId` and the `context` part of the key, to clarify they have nothing to do with this:

> I would say - most developers are working and translating on ONE application at the same time.

These constraints are not for developers, they are for LSPs (language service providers, also known as translators). 

Especially if you are a smaller client, they will most likely leverage their translation memory, which means if you use short keys, they might set the configuration in their tools to generate [ICE matches](https://en.wikipedia.org/wiki/Translation_memory). And in these cases, they might just take whichever translation was in their tool without verifying, which might or might not cause quality issues. By having unique keys you make sure that you never run into these issues and also guaranteed that your company/project will be able to scale without running into these issues.

### 2) Key constraints 

> Your key is required to be only alphanumeric and numbers from 1-50 chars.

I don't think this is the problem. In fact, having large keys are known to cause problems for some translation tools. I think what you are looking for is to be able to use the sentence in the code instead of keys? If so, I had a backlog item for this but since you were the only person asking for this I never prioritized it yet. I also think that having a fallback option could be feasible. If you check the [TODO](https://github.com/Avansai/next-multilingual/blob/main/TODO.md), you can see they are on the list at the bottom.

If the other benefit you are seeing is for the translation itself, this would probably only apply if your developers are doing the translation so that they can see the source text next to the translated text. You could achieve similar results with .properties files by simply opening them side to side. Even IntelliJ (the Jetbrain IDE) have a table view for .properties file which mimics a bit the tools translators are using.

Of course, translators don't need this because the tools they use already provide them with that side-by-side view.

### 3) File type

> Having keys in a full text manner, then also or even better YAML file type can handle the translation files

You could do the same with `.properties` file. You can have spaces in your keys and newlines just like .yaml. But the big question is will translation tools support this?

Remember, `next-multilingual` is not just for developers. In the sense that a lot of i18n tools out there have been developed by developers, for developers. Often by people who have no experience working with LSPs or translators.. which basically makes those tools non-scalable and inadequate for translation services.

**Comment by TomFreudenberg** [comment]
One Sentence her:

> You can have spaces in your keys and newlines just like .yaml.

If I remember correctly, those keys must be written always in paranthesis (") ?

And you can't (if I am correct) enter your value in multiple lines which get conjuncted by the yaml parser

See my examples from above YAML examples.

---

All I'm discussing here is to make a developer's job easier ;-)

**Comment by nbouvrette** [comment]
You can check all the weird ways you can use properties files here (don't rely on the highlight that much because even the Git parser does not follow the official Java standard...):

https://github.com/Avansai/properties-file/blob/main/assets/tests/test-all.properties

Having said this, this is one of the reasons I wrote this new package because the specification for `.properties` files are not very well defined. It's mostly a Java creation and they never really release clear specifications, so a lot of other libraries out there will provide different results when parsing the files.

`.yaml` is probably worse, especially for translation tools because it's a less mature format and it is also much more complex than .properties file - so it's almost certain that it will cause more tooling-related issues.

And yes I know you want to improve the developer experience, we all do. But this is also why `next-multilingual` exists, to bridge the gaps between developers and translators.

Regarding not using a fork, totally make sense but I don't know which comprise you are willing to make or not.

On my end, we have items that have been in the backlog for a while, and to be completely transparent I am a bit more concerned about Next.js 13 right now than any features on the backlog.

I did some tests yesterday and it looks like the future of i18n for Next.js will be quite different than it is today... meaning if we want to continue future support with `next-multilingual` we will have a huge effort ahead (most likely weeks or even months of development) to:

- define how to configure, and get locales across applications when using the `app` directory
- rebuild the new router to support the new `app` directory model
- support SWC and Rust and get rid of Babel once and for all?

**Comment by TomFreudenberg** [comment]
Hey Nicolas

yes the big question ... how to proceed on next 13 is the main point

As I got it now, "app" is still beta and therefore until next year I am pretty sure.

In addition, pages will still be supported and also other's discussing i18n already

https://github.com/vercel/next.js/issues/41980

But yeah, it is something like reinventing the wheel.

Also SWC is more and more important, Babel will have it's counted days until that should be dropped.

Nevertheless, to fit our needs as well, we need to get the optional path as written above.

Not sure if you want to make next-multilingual the next standard for i18n but for that you have to catch a lot of developer's.

Do you have an idea why currently not more are using it?

What about having an additional call to discuss again? I am working currently on a template for a nextjs app starter, so my time is also counted.

Good night
cheers, Tom

**Comment by nbouvrette** [comment]
I'm currently making 0 noise around this repo :) I'm actually surprised it has so many stars.

The main reason I did this repo was for myself... and I am happy to open-source it but one thing I don't want this to become is another i18n solution made by developers, for developers. It would go against everything I learned over the past few years and wouldn't be something I would be proud of maintaining. It's also why in the description I call this an opinionated package because I know it won't be the most popular package and honestly I'm pretty OK with this. I'm not looking for glory or attention :)

Having said this, I was thinking about your situation, correct me if I am wrong but you don't have many benefits from localized URLs? If this is a trade-off you would consider, maybe simply using https://github.com/Avansai/messages-modules would be a good option for you.

A benefit of using this directly is that you can create your own wrapper, which includes the file type and everything you don't like about `next-multilingual`. Of course, you would need to create your own wrapper, but I think it can be done with a few hours of development at most which will probably be better than maintaining a fork.

Let me know if you think this could be a good option and I am happy to continue to help you if you go in this direction.

**Comment by TomFreudenberg** [comment]
Hi Nicolas

>  It's also why in the description I call this an opinionated package because I know it won't be the most popular package and honestly I'm pretty OK with this. I'm not looking for glory or attention :)

Got that and I am fine on that. But let me say - it's a petty because the concept is a good one and more should use it or have the option to use it. I like it very much to have the translation files directly beneath the sources !!!

> you don't have many benefits from localized URLs?

Absolutely correct - we have NONE because we don't use them. The URL is still - based on the app and customer - native mostly german or international and english.

Non of the apps we make is SEO important. Most of the stuff is behind the login wall. So all we are looking for is a nice developer experience when working with translations and translation files.

> Let me know if you think this could be a good option and I am happy to continue to help you if you go in this direction.

Agree on that, I have already thought about in same manner you do.

But at least - isn't it still a kind of fork (downstripped) of next-multilingual with another file type provider and some other opinionated (my) defaults?

**Comment by nbouvrette** [comment]
> Got that and I am fine on that. But let me say - it's a pity because the concept is a good one and more should use it or have the option to use it. I like it very much to have the translation files directly beneath the sources !!!

Thanks. I think the "message modules" concept is probably the most generic concept of `next-multilingual` and also why I created a separate repo just for this. I think in general people tend to have a lot of different opinions when it comes to i18n and that probably it's one of the reasons why Next.js decided to stop their support

> But at least - isn't it still a kind of fork (downstripped) of next-multilingual with another file type provider and some other opinionated (my) defaults?

If you use the message module repo directly you can decide on the file type and rules you want to implement around your message files. You can inspire yourself from the `next-multilingual` implementation here https://github.com/Avansai/next-multilingual/tree/main/src/messages. Probably the only "advanced" feature you would want to keep would be the JSX formatter which you would need copy and then maintain but it should not be a lot of code to manage on your end I would think.

Ultimately if you do this, you own the code.. other than the Babel plugin which does most of the "message module" logic.. so for me, it would no longer be a fork. It's basically your own package.

**Comment by TomFreudenberg** [comment]
Hi Nicolas

ok, so I try to go with messages-modules - If looking for SWC as babel replacement - where would you start? Is this the messages-modules then which has to be changed or is the babel plugin still inside the next-multilingual or "my component"?

Thanks for your open mind and words

**Comment by nbouvrette** [comment]
I originally created the [`messages-modules`](https://github.com/Avansai/messages-modules) package to be able to handle both Babel and SWC plugins.

Unfortunately, it looks like the SWC JavaScript plugins are being replaced by Rust plugins (for performance reasons). The way I originally split `messages-modules` was basically to handle all the plugin logic here:

- https://github.com/Avansai/messages-modules/blob/main/src/index.ts

And then you could use the package to create your own plugin on top (a wrapper) that would be responsible to parse the file and send the key/values to the `messages-modules` plugin package. This is an example of a wrapper:

- https://github.com/Avansai/messages-modules/blob/main/src/babel-plugin/index.ts (not to be used for production)

You can see you could easily replace .properties file by .yaml or whichever else file you want to support. Of course, the example plugin does not include any formatting feature like ICU syntax which is something we added on top of JSX injection in `next-multilingual` here:

- https://github.com/Avansai/next-multilingual/tree/main/src/messages

So now the challenge, is because of the way the boundaries are defined it won't work with Rust.. we would need to write both the `messages-modules` in Rust as well its wrappers. This comes with challenges in terms of being able to use the same file parser between the JavaScript application and the Rust plugin. This might not be a big issue for you for for `next-multilingual` it was nice to be able to parse the files the same way for both the localized URLs and the messages.

Now I don't know if this context helped or not but if you want to dig more into this, maybe there is an opportunity to work together and integrate your work in `messages-modules` since it was something I had on my personal backlog.

Keep me posted!

**Comment by TomFreudenberg** [comment]
Hi Nicolas

> maybe there is an opportunity to work together and integrate your work in messages-modules since it was something I had on my personal backlog.

That's also what I would like to do - keeping our connection und our work together.

I will checkout the messages-modules and RUST opportunities and come back to you.

A question on this: why is "properties-file" a "Dependency" but not only a "devDependency" of messages-modules?

Cheers
Tom

**Comment by nbouvrette** [comment]
I didn't put it as a dev dependency because the package offers a working example plug-in which is based on properties file. This is something we could reconsider but unless you use the example plugin you should never need to use this dependency. It's sort of a convenience dependency that is never really used other than to test/learn to use the package
